utils/src/bitmap/mod.rs

```diff
@@ -0,0 +1,1705 @@
+            next_bit: size % Self::CHUNK_SIZE_BITS,
+        }
+    }
+
+    /// Create a new bitmap with `size` bits, with all bits set to 1.
+    pub fn ones(size: u64) -> Self {
+        let num_chunks = Self::num_chunks_at_size(size).get();
+        let mut chunks = VecDeque::with_capacity(num_chunks);
+        for _ in 0..num_chunks {
+            chunks.push_back(Self::FULL_CHUNK);
+        }
+        Self {
+            chunks,
+            next_bit: size % Self::CHUNK_SIZE_BITS,
+        }
+    }
+
+    /// Return the number of bits currently stored in the bitmap.
+    #[inline]
+    pub fn len(&self) -> u64 {
+        (self.chunks.len() as u64 - 1) * Self::CHUNK_SIZE_BITS + self.next_bit
+    }
+
+    /// Return the last chunk of the bitmap and its size in bits.
+    /// The size can be 0 (meaning the last chunk is empty).
+    #[inline]
+    fn last_chunk(&self) -> (&[u8; N], u64) {
+        (self.chunks.back().unwrap(), self.next_bit)
+    }
+
+    /// Return the last chunk of the bitmap and its size in bits.
+    /// The size can be 0 (meaning the last chunk is empty).
+    #[inline]
+    fn last_chunk_mut(&mut self) -> &mut [u8; N] {
+        self.chunks.back_mut().unwrap()
+    }
+
+    /// Returns the bitmap chunk containing the specified bit.
+    ///
+    /// # Warning
+    ///
+    /// Panics if the bit doesn't exist.
+    #[inline]
+    fn get_chunk(&self, bit_offset: u64) -> &[u8; N] {
+        assert!(bit_offset < self.len(), "out of bounds: {bit_offset}");
+        &self.chunks[Self::chunk_index(bit_offset)]
+    }
+
+    /// Get the value of a bit.
+    ///
+    /// # Warning
+    ///
+    /// Panics if the bit doesn't exist.
+    #[inline]
+    pub fn get_bit(&self, bit_offset: u64) -> bool {
+        let chunk = self.get_chunk(bit_offset);
+        Self::get_bit_from_chunk(chunk, bit_offset)
+    }
+
+    /// Get the value at the given global `bit_offset` from the `chunk`.
+    #[inline]
+    fn get_bit_from_chunk(chunk: &[u8; N], bit_offset: u64) -> bool {
+        let byte_offset = Self::chunk_byte_offset(bit_offset);
+        let byte = chunk[byte_offset];
+        let mask = Self::chunk_byte_bitmask(bit_offset);
+        (byte & mask) != 0
+    }
+
+    /// Add a single bit to the bitmap.
+    pub fn push(&mut self, bit: bool) {
+        let chunk_byte = (self.next_bit / 8) as usize;
+        let next_bit = self.next_bit;
+        let last_chunk = self.last_chunk_mut();
+        // Set the bit
+        if bit {
+            last_chunk[chunk_byte] |= Self::chunk_byte_bitmask(next_bit);
+        } else {
+            last_chunk[chunk_byte] &= !Self::chunk_byte_bitmask(next_bit);
+        }
+        self.next_bit += 1;
+        assert!(self.next_bit <= Self::CHUNK_SIZE_BITS);
+
+        if self.next_bit == Self::CHUNK_SIZE_BITS {
+            self.prepare_next_chunk();
+        }
+    }
+
+    /// Remove and return the last bit from the bitmap.
+    ///
+    /// # Warning
+    ///
+    /// Panics if the bitmap is empty.
```

Comment:
I think you should probably assert this with a descriptive message rather than letting it panic on the unwrap within `self.last_chunk()` at the bottom of this function

Reply:
Added a bunch of `assert` for nicer error message when we get invalid args